util/include/thread_pool.hpp

```diff
@@ -0,0 +1,123 @@
+
+#include <assertUtils.hpp>
+
+#include <atomic>
+#include <condition_variable>
+#include <future>
+#include <mutex>
+#include <queue>
+#include <thread>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+namespace concord::util {
+
+// A thread pool that supports any callable object with any return type. Returns std::future objects to users.
+class ThreadPool {
+ private:
+  std::size_t nextQueueIndex() { return (queue_counter_++) % task_queues_.size(); }
+  auto& nextQueue() { return task_queues_[nextQueueIndex()]; }
+
+ public:
+  // Starts the thread pool with thread_count > 0 threads.
+  ThreadPool(unsigned int thread_count) : task_queues_{thread_count} {
+    Assert(thread_count > 0);
+    for (auto thread_index = 0u; thread_index < thread_count; ++thread_index) {
+      threads_.emplace_back([this, thread_index]() { loop(thread_index); });
+    }
+  }
+
+  // Starts the thread pool with the maximum number of concurrent threads supported by the implementation.
+  ThreadPool() : ThreadPool{std::thread::hardware_concurrency() ? std::thread::hardware_concurrency() : 1} {}
+
+  // Stops the thread pool. Waits for the currently executing tasks only (will not exhaust the queues).
+  ~ThreadPool() noexcept {
+    stop_ = true;
+    for (auto& q : task_queues_) {
+      q.cv.notify_one();
+    }
+    for (auto& t : threads_) {
+      t.join();
+    }
+  }
+
+ public:
+  // Executes the passed function (or any callable) in a pool thread. Returns a future to the result.
+  template <class F>
+  auto async(F&& func) {
+    using ResultType = std::invoke_result_t<std::decay_t<F>>;
+    auto ptask = std::packaged_task<ResultType()>{std::forward<F>(func)};
+    auto future = ptask.get_future();
+    auto task = GenericTask{[ptask = std::move(ptask)]() mutable { ptask(); }};
+    auto& queue = nextQueue();
+    {
+      auto lock = std::lock_guard{queue.mutex};
+      queue.tasks.push(std::move(task));
+    }
+    queue.cv.notify_one();
+    return future;
+  }
+
+ private:
+  using GenericTask = std::packaged_task<void()>;
+
+  void loop(std::size_t thread_index) noexcept {
+    auto& queue = task_queues_[thread_index];
+    while (true) {
+      auto lock = std::unique_lock{queue.mutex};
+      while (queue.tasks.empty() && !stop_) {
+        queue.cv.wait(lock);
+      }
+      if (stop_) break;
+      Assert(!queue.tasks.empty());
+      auto task = std::move(queue.tasks.front());
+      queue.tasks.pop();
+      lock.unlock();
+      task();
+    }
+  }
+
+ private:
+  // Represents a per-thread task queue.
+  struct TaskQueue {
+    // A queue of tasks for execution.
+    std::queue<GenericTask> tasks;
+
+    // A mutex for the queue.
+    std::mutex mutex;
+
+    // A condition variable to signal the queue.
+    std::condition_variable cv;
+  };
+
+  // Each thread has its own queue - rationale is that doing so will lessen the contention compared to having a single
```

Comment:
On the other hand starvation is very probable.
Personally I'd implement a single queue

Reply:
Good point. Do you mean a case where tasks have different complexities?

I think, in general, we can measure it. Also, we can have a policy to specify if there should be one queue only. Another option is to implement work stealing.

Please let me know what your thoughts are on above.

Reply:
After an offline discussion, I've pushed a commit that implements by using a single queue.